csrc/preseg_passes/allocation_order_inference.cpp

```diff
@@ -15,442 +16,279 @@ namespace nvfuser::preseg_passes {
+  auto mapped_id_iter = mapped_id_vec.begin();
+  auto unmapped_id_iter = unmapped_ids_vec.begin();
+  // initialize new target allocation domain with nullptr
+  std::vector<IterDomain*> target_alloc_domain(
+      target_rfactor_domain.size(), nullptr);
+  for (auto i : c10::irange(target_rfactor_domain.size())) {
+    // sharp-edges 1
+    // preserves non-mapped reduction id in its original position
+    if (target_rfactor_domain[i]->isReduction() &&
+        mapped_id_set.count(target_rfactor_domain[i]) == 0) {
+      target_alloc_domain[i] = target_rfactor_domain[i];
       continue;
     }
-
-    auto iter = alloc_order_map_.find(tv);
-    // stopping propagation when we encounter an entry that does not have an
-    // allocation order. See NOTE: [Allocation Order Inference]
-    if (iter == alloc_order_map_.end()) {
-      return nullptr;
+    // push unmapped ids to outer dimension until it's fully consumed
+    if (unmapped_id_iter != unmapped_ids_vec_end) {
+      target_alloc_domain[i] = *unmapped_id_iter++;
+    } else {
+      // push mapped ids to inner dimension
+      target_alloc_domain[i] = *mapped_id_iter++;
     }
-
-    // skip entry that has an empty allocation order
-    if (iter->second.empty()) {
-      // We still want to ensure that we propagate empty allocation order if
-      // there's no candidate with a non-empty allocation order
-      if (src == nullptr) {
-        src = tv;
+  }
+#else
+  // mapping id between ref's allocation domain to target's rfactor domain
+  for (auto* ref_id : ref_alloc_domain) {
+    for (auto* id : target_rfactor_domain) {
+      if (val_sets.permissiveAreMapped(ref_id, id)) {
+        mapped_id_vec.push_back(id);
+        mapped_id_set.insert(id);
+        break;
       }
-
-      // skip if unspecified
-      continue;
-    }
-
-    // check if current entry sets new record for num of non broadcast / non
-    // reduction iterdomain
-    if (size_t non_bc_count = countLoopIterDomains(tv);
-        non_bc_count > non_bc_high_water_mark) {
-      non_bc_high_water_mark = non_bc_count;
-      src = tv;
     }
   }
-
-  return src;
-}
-
-// FullOp set empty allocation order to output
-void AllocationOrderInferencer::handle(FullOp* op) {
-  auto* out = static_cast<TensorView*>(op->output(0));
-  alloc_order_map_[out] = {};
-}
-
-// UnaryOp propagation forward allocation order from input to output
-void AllocationOrderInferencer::handle(UnaryOp* op) {
-  auto* out = dynamic_cast<TensorView*>(op->out());
-  if (out == nullptr) {
-    return;
+  std::vector<IterDomain*> target_alloc_domain = target_rfactor_domain;
+  // removing mapped ids.
+  auto unmapped_ids_vec_end = std::remove_if(
+      target_alloc_domain.begin(),
+      target_alloc_domain.end(),
+      [&mapped_id_set](IterDomain* it) {
+        return mapped_id_set.count(it) != 0;
+      });
+  // appending mapped ids at the end of target_alloc_domain.
+  std::copy(mapped_id_vec.begin(), mapped_id_vec.end(), unmapped_ids_vec_end);
```

Comment:
What if we have `reshape`?

Reply:
This is intentionally left out.

iiuc, reshape that does merge/split/resize wouldn't map to anything and those new dimensions would just be pushed to outer dimensions, which might not be the best decision.
The question is also, how are we treating reshape in IdModel? I tried to use `ComputeAtMap`'s INNERMOST mapping, I think could be one of the things for us to use here. where we'll be able to map through a reshape.

i.e. if we have `i0` -> `i0/4, 4`, we'll be mapping `i0` to `4` here, I think that aligns with what we are trying to do here on mapping the order of inner dimensions from source to destination.
How does that sound like? cc'ing @naoyam 

Having said that, I'm a little bit scared about pushing for that at this moment. reduction scheduler is already struggling with keeping up, exposing more propagation at the time might expose more issues. Which we would want to get there but maybe not yet.

Reply:
I think you'd need to reason about merge/split outside IdModel, e.g., https://github.com/NVIDIA/Fuser/blob/2b115a1ec25ada00332d1e7f754dc1a16db66bf2/csrc/preseg_passes/move_split_cat.cpp#L323-L338. And the logic would be similar to https://github.com/NVIDIA/Fuser/blob/2b115a1ec25ada00332d1e7f754dc1a16db66bf2/csrc/alias_analysis.cpp#L163, although it doesn't use IdModel.

Reply:
yeah, that's basically the ValGraph? I'll just need to do find the path between two ValGroups and reason about that.

Reply:
IIUC there's a partial order of IterDomains in the reference tv's allocation domain. We can extend that to a partial order of `ValGroup`s in a straightforward way. We would like to extend that to a partial order on the ValGraph using lexicographic order i.e. each node of the graph orders its children. Then use that ordering to sort the consumer allocation domain. Some rules on the ValGraph would be that `a < c && b < c` implies `Merge(a, b) < c`, `a < b && a < c` implies `a < Merge(b, c)`, and `b, c = Split(a)` implies `b < c`. There are upstream implications implied by the contrapositives and transitivity of `<` as well: `b, c = Split(a) && b < d && c < d` implies `a < d` and `c = Merge(a, b) && c < d` implies `a < d && b < d` for example. Other expressions like `Resize` have no implications; e.g. `b = Resize(a)` implies nothing about `b < c` given either `a < c` or `c < a`. You'd start with the orders implied by the reference tensor and all the split outputs, then you'd complete the partial order by propagating these rules. Does that sound right?

Reply:
I think @jacobhinkle 's reasoning here is that we should try to establish many-to-many mapping and a order of ids in the many-to-many mapping and propagate that through the graph.

That's more than what I was thinking, but sounds like a reasonable suggestion as well.

Reply:
Recording the alternatives in issue #2235 .